### ISSUE TITLE
DataArray.unstack() leaving dimensions 'in order'

### ISSUE THREAD

**Issue by takluyver** [pydata/xarray#3786]
It's documented that when you `.unstack()` a DataArray, "New dimensions will be added at the end."

I'm using `.unstack()` kind of like `np.reshape()`, to split one dimension into two. From this perspective, it's surprising that new dimensions go at the end, rather than in place of the old dimension. This also means that either the new array is no longer C-contiguous, or the `.unstack()` operation has had to copy all the data to rearrange it. It appears I'm not the only one to think this: the code sample in #2419 (about adding a reshape method) does the same workaround that I do: unstack and then carefully reorder the dimensions.

First, I'd like to understand: why the end? It seems like there's an obvious place to put unstacked dimensions (`ab, c -> a, b, c`), but presumably there's some reason it doesn't work that way.

Then, is there anything one can do with xarray today to 'unstack' a dimension while keeping the order, short of using `.transpose()` afterwards to recreate the desired order? And if not, is there any potential to add a feature for that?

I guess this might touch on missing data: `unstack()` can create and handle 'gaps', whereas `reshape()` can't. If so, is there some more limited operation that could only unstack a perfectly regular multiindex - where it's guaranteed to separate into 2 dimensions without creating gaps?

**Comment by max-sixty** [comment]
Thanks for the issue @takluyver 

I mostly think (others should weigh in) that we don't spend much effort on dimension order. (Would a dimension order change always be a regression?). That's mostly because of one of xarray's great strengths: you don't need to worry about the order, because you can reference dimensions by name.

I imagine that `unstack` dimension order wasn't carefully chosen to be at the end vs in place; and we'd take a PR to make that change.

**Comment by takluyver** [comment]
Thanks, it makes sense that dimension order conceptually doesn't matter so much once they're labelled. Though I'd say that as Xarray has public APIs for accessing dimensions by position, changing how something like `.unstack()` orders them probably is a breaking change.

I'm providing an API for other people to access data as a DataArray, and I'd like to have the dimensions in the order that makes it C contiguous, as a hint about what kinds of operations will be efficient. I know some users also go 'what is this complicated thing?' and extract the numpy array from it, in which case the order is more important.

So another option that would work for me would be a separate method that rearranges the dimensions to the order that makes the array C contiguous.

**Comment by crusaderky** [comment]
> This also means that either the new array is no longer C-contiguous, or the .unstack() operation has had to copy all the data to rearrange it. 

The former. As a core design principle, xarray does not care about dimensions order, and any user code that implicitly relies on it should be considered a bad design.
The .transpose() method mostly only exists for when people need to access the numpy .data object directly with a numpy function.

**Comment by max-sixty** [comment]
So I'm understanding correctly: does numpy's behavior even satisfy your reqs? https://docs.scipy.org/doc/numpy/reference/generated/numpy.swapaxes.html

> then a view of a is returned

i.e. the array's underlying layout doesn't change, only the view...

**Comment by takluyver** [comment]
@max-sixty - I certainly want to avoid copying the data unless it's necessary. But I'd like to present the axes in the 'real' memory order, from the largest stride to the smallest.

I appreciate that it shouldn't matter for program logic, but it can definitely matter for performance, and I know some users are going to use axes by position rather than by name, so I do consider it an important part of the API.

**Comment by max-sixty** [comment]
I think I see your point @takluyver — that by returning the new axes at the end it's both surprising _and_ no longer a contiguous array.

But then +1 re your final point about unstack being unable to return a view on missing data.

I don't think it's impossible to carefully manage the dimensions order & contiguous-ness in this case. It may take someone who can champion it, though

**Comment by takluyver** [comment]
If it's easier, a method for "transpose these dimensions to the order which makes the data contiguous" would meet my needs.

I'm happy in principle to work on either feature, but when I've looked into contributing to xarray before, I've been a bit overwhelmed by complexity - I think I'm currently using a pretty small fragment of what it can do.

**Comment by max-sixty** [comment]
I think it should be pretty reasonable to do that in a few lines, for numpy-backed arrays at least! Get the underlying numpy array, sort the `.strides` to get the memory order, map to dimension names, and pass them to `transpose`?

**Comment by dcherian** [comment]
This sounds useful to me

**Comment by takluyver** [comment]
Belatedly coming back to this: I think outright changing the order of dimensions that unstack returns would be an API break. Even if referring to dimensions by name is preferred, there are plenty of public & well documented APIs to access them positionally.

The transpose-to-memory-layout operation doesn't entirely do what I want here, because depending on the code path taken, unstack actually copies the data to a new array with the 'wrong' memory layout ([even when that's avoidable](https://github.com/pydata/xarray/issues/11455)). I don't want the dimension order to depend on an implementation detail.

Would an option like `arr.unstack(preserve_order=True)` be acceptable? I.e. with preserve order, dimensions `ab, c` -> `a, b, c`, and without it `ab, c` -> `c, a, b` (the status quo).

**Comment by dcherian** [comment]
We have another way of expressing a reshape in `coarsen.construct`. Does that work for you?

**Comment by takluyver** [comment]
`coarsen.construct` seems to do what I want with the dimension order, and it's more efficient than unstack on this data where no fill values are needed. But it doesn't appear to know about the coordinates on the dimension that I want to unstack, so I need to find the length of the repeating group, possibly verify that it is repeating, and then fix up the coordinates afterwards. I've got the code to do all of this, just with a numpy reshape call in the middle of it. 🙂

**Comment by dcherian** [comment]
>  But it doesn't appear to know about the coordinates on the dimension that I want to unstack, so I need to find the length of the repeating group, possibly verify that it is repeating, and then fix up the coordinates afterwards. 

I didn't understand this but yes, coarsen is indeed "uniform windows" and does a reshape under the hood. Presumably your coordinates are out of order and you need a sort?

**Comment by takluyver** [comment]
The data has a MultiIndex where the inner level is a repeating pattern. To make it concrete, the data is organised in 'pulse trains', where each train contains 10s - 100s of pulses.

<img width="730" height="182" alt="Image" src="https://github.com/user-attachments/assets/124403ed-4aec-4f5f-8bae-16f9d6c6a51d" />

Using unstack to split this into 2 dimensions, each new dimension gets a simple index:

<img width="726" height="150" alt="Image" src="https://github.com/user-attachments/assets/5f45625e-e55d-4a78-a85b-69c10e565225" />

Using `.coarsen().construct()` as a workaround, it keeps the original index as 2D coordinates - as I imagine it has to, since it doesn't 'know' that the index is regular. But this is what I mean about 'fixing up' coordinates - getting from here to the coordinates unstack gives.

<img width="722" height="181" alt="Image" src="https://github.com/user-attachments/assets/d307134b-a679-43c4-b97c-3ef42620ee40" />

**Comment by dcherian** [comment]
Yes the 'xarray' way of indicating that orthogonality is a groupby. We do not have `groupby.construct` but we could. The "ordered groups" case (we call this "resample") that you have admits the fast `np.reshape` path for same-sized groups. 

I prototyped something with Claude [here](https://github.com/pydata/xarray/discussions/10919#discussioncomment-14974152) for the more general case. So we could start with that and then optimize it.

Actually using `resample` is a bit complicated; you'll need to create a [custom "Resampler"](https://docs.xarray.dev/en/stable/user-guide/groupby.html#grouper-objects) which is totally doable (perhaps there are other ways a PulseTrainResample` would be useful?). But it should be easy to detect the ordered-groups case, check for same-sized-ness, and use `np.reshape`. That way we can use the standard `groupby(["trainId", "pulseId"])` path.

**Comment by dcherian** [comment]
That said, i have felt this particular pain with `coarsen.construct` before... I couldn't come up with an API for the user to indicate invariance per dimension per variable.
